### PR TITLE
Resize relies on interpolate's native uint8 handling

### PR DESCRIPTION
**Pull request by vfdev-5**
Description:

- Now that https://github.com/pytorch/pytorch/pull/90771 is merged, let Resize() rely on interpolate()'s native uint8 handling instead of converting to and from float.

  - uint8 input is not casted to f32 for nearest mode and bilinear mode if the latter has AVX2.

Context: https://github.com/pytorch/vision/issues/7217

Benchmarks:
```
[----------- Resize cpu torch.uint8 InterpolationMode.NEAREST -----------]
                         |  resize v2  |  resize stable  |  resize nightly
1 threads: ---------------------------------------------------------------
      (3, 400, 400)      |      457    |        461      |        480
      (16, 3, 400, 400)  |     6870    |       6850      |      10100

Times are in microseconds (us).

[---------- Resize cpu torch.uint8 InterpolationMode.BILINEAR -----------]
                         |  resize v2  |  resize stable  |  resize nightly
1 threads: ---------------------------------------------------------------
      (3, 400, 400)      |      326    |        329      |        844
      (16, 3, 400, 400)  |     4380    |       4390      |      14800

Times are in microseconds (us).
```

[Source](https://gist.github.com/vfdev-5/a2e30ed50b5996807c9b09d5d33d8bc2)

### CODE REVIEW

File: torchvision/transforms/v2/functional/_geometry.py [hunk 1]
```diff
@@ -183,7 +183,19 @@ def resize_image_tensor(
...
-        need_cast = dtype not in (torch.float32, torch.float64)
+        acceptable_dtypes = [torch.float32, torch.float64]
+        if interpolation in [InterpolationMode.NEAREST, InterpolationMode.NEAREST_EXACT]:
+            # uint8 dtype can be included for cpu and cuda input if nearest mode
+            acceptable_dtypes.append(torch.uint8)
+        elif interpolation == InterpolationMode.BILINEAR and image.device.type == "cpu":
+            # uint8 dtype support for bilinear mode is limited to cpu and
+            # according to our benchmarks non-AVX CPUs should prefer u8->f32->interpolate->u8 path
+            # TODO: enable torchscript and torch.backends.cpu.get_cpu_capability
+            if not torch.jit.is_scripting():
```
[the remark this answers is not in the document]
Comment (NicolasHug) [reply torchvision/transforms/v2/functional/_geometry.py:194]: Scratch that. My  comment above only applies to the transform classes, but we still case about the torchscriptability of the v2 functionals

File: torchvision/transforms/v2/functional/_geometry.py [hunk 2]
```diff
@@ -185,7 +185,35 @@ def resize_image_tensor(
...
+                    image = image.contiguous(memory_format=torch.channels_last)
+
+        if image.is_contiguous(memory_format=torch.channels_last):
+            strides = image.stride()
+            numel = image.numel()
+            if image.shape[0] == 1 and numel != strides[0]:
+                # This is the case when channels last tensor was squeezed and unsqueezed such that
+                # stride[0] set as image.shape[1] * image.stride()[1] instead of being image.numel()
+                # Let's restride image such that it will be correctly treated as channels last.
+                # Related pytorch issue: https://github.com/pytorch/pytorch/issues/68430
```
Comment (NicolasHug) [inline torchvision/transforms/v2/functional/_geometry.py:211]: I think here it's important to provide details of what we're tying to avoid (an extra copy of the output), and when we should be able to remove those checks. May I suggest the following:

> There is a weird behaviour in torch core where the output tensor of `interpolate()` can be allocated as contiguous even though the input is un-ambiguously channels_last (https://github.com/pytorch/pytorch/issues/68430). In particular this happens for the typical torchvision use-case of single CHW images where we fake the batch dim to become 1CHW. Below, we restride those tensors to trick torch core into properly allocating the output as channels_last, thus preserving the memory format of the input. This is not just for format consistency: for uint8 bilinear images, this also avoids an extra copy (re-packing) of the output and saves time.
TODO: when https://github.com/pytorch/pytorch/issues/68430 is fixed (possibly by https://github.com/pytorch/pytorch/pull/100373), we should be able to remove this hack.

File: test/common_utils.py [hunk 3]
```diff
@@ -550,19 +559,26 @@ def fn(shape, dtype, device):
...
         )
 
-        image_tensor = convert_dtype_image_tensor(to_image_tensor(image_pil).to(device=device), dtype=dtype)
+        image_tensor = to_image_tensor(image_pil)
+        if memory_format == torch.contiguous_format:
+            image_tensor = image_tensor.to(device=device, memory_format=memory_format, copy=True)
+        else:
+            image_tensor = image_tensor.to(device=device)
+        image_tensor = convert_dtype_image_tensor(image_tensor, dtype=dtype)
+        assert image_tensor[None].is_contiguous(memory_format=memory_format)
```
Comment (NicolasHug) [inline test/common_utils.py:568]: do we really want this assert here, or was it just a sanity check during development?

File: test/test_transforms_v2_consistency.py [hunk 4]
```diff
@@ -98,6 +98,8 @@ def __init__(
             ArgsKwargs((29, 32), antialias=False),
             ArgsKwargs((28, 31), antialias=True),
         ],
+        # rtol=atol=1 due to Resize v2 is using native uint8 interpolate path for biliear and nearest modes
```
Comment (NicolasHug) [inline test/test_transforms_v2_consistency.py:101]: nit (same below)

```suggestion
        # rtol=atol=1 due to Resize v2 is using native uint8 interpolate path for bilinear and nearest modes
```

File: test/test_transforms_v2_consistency.py [hunk 5]
```diff
@@ -98,6 +98,8 @@ def __init__(
...
+        closeness_kwargs=dict(rtol=1, atol=1),
```
Comment (NicolasHug) [inline test/test_transforms_v2_consistency.py:102]: shouldn't this be `rtol=0, atol=1`?

File: test/test_transforms_v2_functional.py [hunk 6]
```diff
@@ -1365,3 +1365,28 @@ def test_correctness_uniform_temporal_subsample(device):
...
+)
+def test_memory_format_consistency_resize_image_tensor(test_id, info, args_kwargs):
+    (input, *other_args), kwargs = args_kwargs.load("cpu")
+
+    output = info.kernel(input.as_subclass(torch.Tensor), *other_args, **kwargs)
+
+    error_msg_fn = parametrized_error_message(input, *other_args, **kwargs)
+    assert input.ndim == 3, error_msg_fn
+    input_stride = input.stride()
+    output_stride = output.stride()
```
Comment (NicolasHug) [inline test/test_transforms_v2_functional.py:1384]: Would it be simpler to check that `output_mem_format == input_mem_format` ?

**Review by NicolasHug — COMMENTED** [review]
Thanks @vfdev-5 , PR looks good. I made a few comments, LMK what you think

File: test/common_utils.py (hunk 3 above)
Comment (vfdev-5) [reply test/common_utils.py:568]: Yeah, a sanity check to ensure that input is the one we want. I can remove it you think it is unnecessary.

File: test/test_transforms_v2_functional.py (hunk 6 above)
Comment (vfdev-5) [reply test/test_transforms_v2_functional.py:1384]: It is a temporary test before fixing https://github.com/pytorch/pytorch/issues/68430 and merging a PR related to that. Also, here tensors are 3D so there is no direct way to get if a tensor is CL. We have to make the following : 
```
t_aux = t if t.ndim == 4 else t[None]
mf = torch.contiguous_format if t_aux.is_contiguous() else torch.channels_last
```
This is also ugly and bulky

File: torchvision/transforms/v2/functional/_geometry.py [hunk 7]
```diff
@@ -185,7 +185,35 @@ def resize_image_tensor(
...
+                # TODO: Remove when https://github.com/pytorch/pytorch/pull/101136 is landed
+                if dtype == torch.uint8 and not (
+                    image.is_contiguous() or image.is_contiguous(memory_format=torch.channels_last)
+                ):
+                    image = image.contiguous(memory_format=torch.channels_last)
+
+        if image.is_contiguous(memory_format=torch.channels_last):
+            strides = image.stride()
+            numel = image.numel()
+            if image.shape[0] == 1 and numel != strides[0]:
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply torchvision/transforms/v2/functional/_geometry.py:207]: We reshape the input into a 4D tensor, line 185:
```
image = image.reshape(-1, num_channels, old_height, old_width)
```
https://github.com/pytorch/vision/pull/7557/files#diff-cc7bbc66ec0ae77a2f63080195693233155d8ba6b81df18d159bf29f5bc18f34R185 

So, I think there is no need to recheck if 4D

File: test/test_transforms_v2_functional.py (hunk 6 above)
Comment (NicolasHug) [reply test/test_transforms_v2_functional.py:1384]: fair enough - some comments may help to indicate what exactly we're trying to check with those strides. For example it's worth clarifying what the 2 cases below correspond to (`input_stride[-1] == 1` and  `input_stride[0] == 1`)

File: test/test_transforms_v2_functional.py [hunk 8]
```diff
@@ -1365,3 +1365,28 @@ def test_correctness_uniform_temporal_subsample(device):
...
+    if input_stride[-1] == 1:
+        expected_stride = (output.shape[-2] * output.shape[-1], output.shape[-1], 1)
+        assert expected_stride == output_stride, error_msg_fn("")
```
Comment (NicolasHug) [inline test/test_transforms_v2_functional.py:1387]: curious why we need `error_msg_fn` here and can't just use a plain `assert` as usual?

Comment (vfdev-5) [reply test/test_transforms_v2_functional.py:1387]: error_msg_fn reports more info about the tested input

File: test/test_transforms_v2_consistency.py (hunk 5 above)
Comment (vfdev-5) [reply test/test_transforms_v2_consistency.py:102]: There is a case with the following error if run locally:
```
FAILED test/test_transforms_v2_consistency.py::test_call_consistency[Resize-12]

Mismatched elements: 382 / 11136 (3.4%)
Greatest absolute difference: 103 at index (1, 1, 14, 19) (up to 1 allowed)
Greatest relative difference: 1.0 at index (0, 0, 14, 10) (up to 0 allowed)
```

**Review by NicolasHug — APPROVED** [review]
Alright, tests are passing and we also stress-tested the changes locally  against the float implementation. Looks safe to merge now.

Thanks a ton @vfdev-5 for all your work on `interpolate()` and with bug chasing / fixing!

<details> 

```py
import random

import torch
from torchvision.transforms import Resize as Resize_v1
from torchvision.transforms.v2 import Resize as Resize_v2
import pytest

@pytest.mark.parametrize("C", range(1, 6))
@pytest.mark.parametrize("batch_size", ("3d", 1, 4))
@pytest.mark.parametrize("memory_format", (torch.contiguous_format, torch.channels_last, "strided", "cropped"))
@pytest.mark.parametrize("antialias", (True, False))
@pytest.mark.parametrize("seed", range(100))
def test_resize(C, batch_size, memory_format, antialias, seed):
    
    torch.manual_seed(seed)
    random.seed(seed)

    Hi = 2**random.randint(3, 10) + random.randint(0, 30) 
    Wi = 2**random.randint(3, 10) + random.randint(0, 30) 
    Ho = 2**random.randint(3, 10) + random.randint(0, 30) 
    Wo = 2**random.randint(3, 10) + random.randint(0, 30) 
    print(Hi, Wi, Ho, Wo)

    if batch_size == "3d":
        img = torch.randint(0, 256, size=(C, Hi, Wi), dtype=torch.uint8)
    else:
        img = torch.randint(0, 256, size=(batch_size, C, Hi, Wi), dtype=torch.uint8)

    if memory_format in (torch.contiguous_format, torch.channels_last):
        if batch_size == "3d":
            return
        img = img.to(memory_format=memory_format, copy=True)
    elif memory_format == "strided":
        if batch_size == "3d":
            img = img[:, ::2, ::2]
        else:
            img = img[:, :, ::2, ::2]
    elif memory_format == "cropped":
        a = random.randint(1, Hi // 2)
        b = random.randint(Hi // 2 + 1, Hi)
        c = random.randint(1, Wi // 2)
        d = random.randint(Wi // 2 + 1, Wi)
        if batch_size == "3d":
            img = img[:, a:b, c:d]
        else:
            img = img[:, :, a:b, c:d]
    else:
        raise ValueError("Uh?")
    

    out_uint8 = Resize_v2(size=(Ho, Wo), antialias=antialias)(img)
    out_float = Resize_v1(size=(Ho, Wo), antialias=antialias)(img.float()).round().to(torch.uint8)

    torch.testing.assert_close(out_uint8, out_float, rtol=0, atol=1)

        
```
</details>
